Handle constant SymBool in unary and binary operations

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109171
* #109170
* __->__ #109169

In this PR:
- When Constant SymNode are detected in unary/binary ops demote them to plain int/bool before proceeding. Sometimes this means doing a unary op with a Constant SymNode would result in a plain bool.
- Introduce an is_symbolic method, only available from Python. We need this because isinstance(x, SymInt) is no longer sufficient to check whether a given int/SymInt is symbolic or not. See later PR in the stack to see how this is used.

soulitzer committed:
Support unary and binary operations on constant SymBool

[ghstack-poisoned]

soulitzer committed:
Update on "Support unary and binary operations on constant SymBool"


In this PR:
- Introduce an is_symbolic method, only available from Python. We need this because isinstance(x, SymInt) is no longer sufficient to check whether a given int/SymInt is symbolic or not. See later PR in the stack to see how this is used.
- When Constant SymNode are detected in unary/binary ops demote them to plain int/bool before proceeding. Sometimes this means doing a unary op with a Constant SymNode would result in a plain bool.

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1476,16 +1489,47 @@ def _make_user_magic(method, user_type):
 
     def binary_magic_impl(self, other):
+        if is_constant(self):
+            return (method_to_operator(method))(get_constant(self), other)
```
Does this handle other not being a constant correctly? 🤔

soulitzer replied:
Yeah, I think so - but let me know if you think I should test this case more deeply (we already do have a test, it just checks if the output is symbolic and guard_bool on that returns the expected value). What should happen is:
- If other is a symbolic SymBool, calling the operator with op(plain bool, symbolic SymBool) redispatches to the reflected op.
- At that point neither argument is constant and we proceed as usual (e.g. with symbolic SymBool promoting the "other" to another symbolic SymBool)

ezyang replied:
oh you have a comment on this

ezyang commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1253,9 +1253,27 @@ void initJITBindings(PyObject* module) {
+            return (
+                !node->singleton_int().has_value()
+                && !node->constant_bool().has_value()
+                && !node->constant_int().has_value()
```
It would be more symmetric with the other functions to add a new virtual method on SymNode and call that directly to query this information

soulitzer replied:
Done.

soulitzer committed:
Update on "Support unary and binary operations on constant SymBool"


In this PR:
- When Constant SymNode are detected in unary/binary ops demote them to plain int/bool before proceeding. Sometimes this means doing a unary op with a Constant SymNode would result in a plain bool.
- Introduce an is_symbolic method, only available from Python. We need this because isinstance(x, SymInt) is no longer sufficient to check whether a given int/SymInt is symbolic or not. See later PR in the stack to see how this is used.

[ghstack-poisoned]

soulitzer commented on test/test_dynamic_shapes.py:
```diff
@@ -714,6 +715,73 @@ def test_method(self, fn, first_type, second_type):
+        test_constant_symbool(j1 != j3, True)
+        test_constant_symbool(j1 != j2, False)
+
+        x = j1 == j2
```
Yes, we can! It is updated now to do so. Originally I had preferred handling constant SymBools in the inputs vs the inputs, since I assumed that there are other ways to get constant SymBools to appear as inputs, but after looking again, it seems unlikely to see them in the wild anymore. I'm considering to remove all the logic to handle constant SymBools in the inputs once I make sure that they don't appear for more compilation test cases. But for now, I've added an alternative way of testing this logic that we currently have.

soulitzer committed:
Update on "Handle constant SymBool in unary and binary operations"


In this PR:
- When Constant SymNode are detected in unary/binary ops demote them to plain int/bool before proceeding. Sometimes this means doing a unary op with a Constant SymNode would result in a plain bool.
- Introduce an is_symbolic method, only available from Python. We need this because isinstance(x, SymInt) is no longer sufficient to check whether a given int/SymInt is symbolic or not. See later PR in the stack to see how this is used.

[ghstack-poisoned]
